crypto/cms/cms_sd.c

```diff
@@ -725,8 +727,13 @@ static int cms_SignerInfo_content_sign(CMS_ContentInfo *cms,
         unsigned char md[EVP_MAX_MD_SIZE];
         unsigned int mdlen;
 
-        if (!EVP_DigestFinal_ex(mctx, md, &mdlen))
-            goto err;
+        if (precomp_md != NULL) {
+            memcpy(md, precomp_md, precomp_mdlen);
```

Comment:
Instead of memcpying (where you would have to also check that the `precomp_mdlen <= sizeof(md))` I'd do it the other way around, I.E. `precomp_md = md` if `precomp_md == NULL`.

Reply:
I was worried about who is responsible for freeing it. With memcpy, the caller is responsible for freeing it afterwards but without memcpy, the caller gives away the responsibility. Is that OK?

Reply:
I was wrong here. No pointer to the data is retained in this case. In the comment "And here as well" below, a pointer to the sig data is retained though.